plrust/src/user_crate/build.rs

```diff
@@ -192,3 +207,39 @@ impl FnBuild {
         &self.crate_dir
     }
 }
+
+/// Asks Postgres, via FFI, for all of its compile-time configuration data.  This is the full
+/// set of things that Postgres' `pg_config` configuration tool can report.
+///
+/// The returned tuple is a `(key, value)` pair of the configuration name and its value.
+fn pg_config_values() -> impl Iterator<Item = (String, String)> {
+    unsafe {
+        // SAFETY:  we know the memory context we're switching to is valid because we're also making
+        // it right here.  We're also responsible for pfreeing the result of `get_configdata()` and
+        // the easiest way to do that is to simply free an entire memory context at once
+        PgMemoryContexts::new("configdata").switch_to(|_| {
+            let mut nprops = 0;
+            // SAFETY:  `get_configdata` needs to know where the "postmaster" executable is located
+            // and `pg_sys::my_exec_path` is that global, which Postgres assigns once early in its
+            // startup process
+            let configdata = pg_sys::get_configdata(pg_sys::my_exec_path.as_ptr(), &mut nprops);
+
+            // SAFETY:  `get_configdata` will never return the NULL pointer
+            let slice = std::slice::from_raw_parts(configdata, nprops);
+
+            let mut values = Vec::with_capacity(nprops);
+            for e in slice {
+                // SAFETY:  the members (we use) in `ConfigData` are properly allocated char pointers,
+                // done so by the `get_configdata()` call above
+                let name = CStr::from_ptr(e.name);
+                let setting = CStr::from_ptr(e.setting);
+                values.push((
+                    name.to_string_lossy().to_string(),
+                    setting.to_string_lossy().to_string(),
+                ))
+            }
+
+            values.into_iter()
```

Comment:
IIUC this memory context will be destroyed when `values.into_iter()` falls out of scope?